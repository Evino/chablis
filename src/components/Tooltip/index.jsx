import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { colors, fontSizes, fontWeights, sizes } from '../../styles/variables';
import { fadeIn } from '../../styles/animations';

const TooltipWrapper = styled.span`
  font-size: ${fontSizes.SMALL};
  font-weight: ${fontWeights.LIGHT};
  padding: ${sizes.HALF} ${sizes.LARGE};
  color: ${colors.WHITE};
  border-radius: ${sizes.RADIUS};
  position: absolute;
  min-width: 255px;
  max-width: 300px;
  text-align: center;
  animation: ${fadeIn} .2s ease-in;
  display: ${props => props.show ? 'block' : 'none'};
  ${props => props.alignY}: calc(100% + 12px);
  ${props => props.alignX}: ${props => props.spaced ? '-55px' : '5px'};

  ${props => props.type === 'default' && css`
    background: ${colors.LIGHTEST_GREY}};
    color: ${colors.DARK_GREY};
  `}

  ${props => props.type === 'success' && css`
    background: ${colors.PRIMARY};

    &:after {
      border-color: ${colors.PRIMARY};
    }
  `}

  ${props => props.type === 'danger' && css`
    background: ${colors.SECONDARY};
  `}

  &:after {
    content: '';
    border-left: 12px solid transparent !important;
    border-right: 12px solid transparent !important;
    border-top: 9px solid;
    width: 0;
    position: absolute;
    ${props => props.alignX}: ${props => props.spaced ? '80px' : '20px'};
    ${props => props.alignY}: -9px;

    ${props => props.alignY === 'top' && css`
      border-top: 0;
      border-bottom: 9px solid;
    `}

    ${props => props.type === 'success' && css`
      border-color: ${colors.PRIMARY};
    `}

    ${props => props.type === 'danger' && css`
      border-color: ${colors.SECONDARY};
    `}

    ${props => props.type === 'default' && css`
      border-color: ${colors.LIGHTEST_GREY};
    `}
  }
`;

export default class Tooltip extends PureComponent {
  static propTypes = {
    text: PropTypes.string.isRequired,
    spaced: PropTypes.bool,
    show: PropTypes.bool,
    element: PropTypes.element,
    align: PropTypes.oneOf([
      'bottom-start',
      'bottom-end',
      'top-start',
      'top-end'
    ]),
    type: PropTypes.oneOf(['default', 'success', 'danger']),
  }

  static defaultProps = {
    show: true,
    align: 'top-end',
    type: 'default',
  }

  getPosition = (align) => {
    switch (align) {
      case 'top-start':
        return ['left', 'bottom'];
      case 'top-end':
        return ['right', 'bottom'];
      case 'bottom-start':
        return ['left', 'top'];
      case 'bottom-end':
      default:
        return ['right', 'top'];
    }
  }

  render() {
    const position = this.getPosition(this.props.align);

    return (
      <TooltipWrapper
        {...this.props}
        alignX={position[0]}
        alignY={position[1]}
      >
        {this.props.text}
      </TooltipWrapper>
    );
  }
}
