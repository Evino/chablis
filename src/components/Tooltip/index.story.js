/* eslint-disable */
import React from 'react';
import { storiesOf } from '@storybook/react';
import Tooltip from './';
import { colors, sizes } from '../../styles/variables';

storiesOf('Tooltip', module)
  .addWithInfo('Default Tooltip', () => (
    <Tooltip
      text="This is a cool tooltip"
      alignX="right"
    />
  ))
  .addWithInfo('Big Tooltip', () => (
    <Tooltip
      text="This is a really big tooltip with an awesome and big text"
      alignX="right"
    />
  ))
  .addWithInfo('Success Tooltip', () => (
    <Tooltip
      text="This is a cool tooltip"
      type="success"
    />
  ))
  .addWithInfo('Danger Tooltip', () => (
    <Tooltip
      text="This is a cool tooltip"
      type="danger"
    />
  ))
  .addWithInfo('Spaced Tooltip', () => (
    <Tooltip
      text="This is a cool tooltip"
      spaced
    />
  ))
  .addWithInfo('Custom aligned Tooltip', () => (
    <div style={{marginTop: '10px'}}>
      <Tooltip
        text="This is a cool tooltip"
        alignX="left"
        alignY="top"
      />
    </div>
  ))

  // <div
  //     style={{
  //       background: colors.LIGHT_GREY,
  //       color: colors.DARKEST_GREY,
  //       display: 'inline-block',
  //       padding: sizes.DEFAULT,
  //       textAlign: 'center',
  //       marginTop: sizes.LARGE,
  //       marginLeft: sizes.LARGE,
  //       position: 'relative',
  //       borderRadius: sizes.RADIUS
  //     }}
  //   >