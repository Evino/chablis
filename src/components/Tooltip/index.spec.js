import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Tooltip from './';

const text = 'This is a tooltip';
const element = <Tooltip text={text} />;
const component = mount(element);

const getPosition = position => component.instance().getPosition(position);

describe('Tooltip', () => {
  it('should render a tooltip', () => {
    const tree = renderer
      .create(element)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should display passed text', () => {
    expect(component.prop('text')).toBe(text);
  });

  it('should return right positions', () => {
    expect(getPosition()).toEqual(['right', 'top']);
    expect(getPosition('top-start')).toEqual(['left', 'bottom']);
    expect(getPosition('top-end')).toEqual(['right', 'bottom']);
    expect(getPosition('bottom-start')).toEqual(['left', 'top']);
    expect(getPosition('bottom-end')).toEqual(['right', 'top']);
  });
});
