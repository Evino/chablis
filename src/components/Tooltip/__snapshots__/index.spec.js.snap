// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Should render a tooltip 1`] = `
<span
  className="sc-bdVaJa bTSTZT"
  type="default"
>
  This is a tooltip
</span>
`;
